Generics ternary operator assignment changes

NikitaAware opened the pull request:
Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [x] A description about what and why you are contributing, even if it's trivial.

  - [x] The issue number(s) or PR number(s) in the description if you are contributing in response to those.

  - [x] If applicable, unit tests.

NikitaAware committed:
tests

NikitaAware committed:
funtions

NikitaAware committed:
changes

msridhar committed:
Merge branch 'generics-return-type-changes' into generics-ternary-operator-assignment-changes

msridhar committed:
Merge branch 'generics-return-type-changes' into generics-ternary-operator-assignment-changes

msridhar committed:
fix

msridhar committed:
change approach

msridhar committed:
add another test case

msridhar committed:
Merge branch 'generics-return-type-changes' into generics-ternary-operator-assignment-changes

msridhar commented:
@NikitaAware I merged the latest return type changes here; please update this PR to work again (with the change in `compareNullabilityAnnotations`) and create a separate error message for conditional expressions

msridhar committed:
Merge branch 'generics-return-type-changes' into generics-ternary-operator-assignment-changes

NikitaAware committed:
updated error message

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -360,4 +387,39 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
     return finalType;
   }
+
+  public void checkTypeParameterNullnessForAssignabilityForConditionalExpression(
```
We should write some javadoc describing what check this method is doing.  Also I think the following is a better name:
```suggestion
  public void checkTypeParameterNullnessForConditionalExpression(
```

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -150,6 +151,32 @@ private static void reportInvalidReturnTypeError(
+      Type rhsType,
+      VisitorState state,
+      NullAway analysis,
+      boolean isTruePartOfTheExpression) {
```
Rather than passing this boolean parameter, we should pass either the true or false part sub-expression as the `tree` argument.  Then the error message can be something like `"conditional expression must have type " + expressionType + ", but sub-expression has type " + rhsType + ", which has mismatched nullability type parameters".  We can get rid of the boolean parameter then

NikitaAware replied:
will update the error message

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -360,4 +387,39 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
+        if (!compareNullabilityAnnotations(
+            (Type.ClassType) condExprType, (Type.ClassType) truePartType)) {
+          reportMismatchedTypeForTernaryOperator(
+              tree, condExprType, truePartType, state, analysis, true);
```
Here pass `truePartTree` instead of `tree`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -360,4 +387,39 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
+        if (!compareNullabilityAnnotations(
+            (Type.ClassType) condExprType, (Type.ClassType) falsePartType)) {
+          reportMismatchedTypeForTernaryOperator(
+              tree, condExprType, falsePartType, state, analysis, false);
```
Here pass `falsePartTree` instead of `tree`

msridhar reviewed (commented):
Looks good, a few comments

NikitaAware committed:
update method name

NikitaAware committed:
javadoc

NikitaAware committed:
updated error message

NikitaAware committed:
updated error message

NikitaAware committed:
suggested changes

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -360,4 +378,45 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
   }
+
+  /**
+   * For the Generic conditional expressions, Check whether the type parameters for the sub-parts of
+   * the conditional statement have the same nullability annotations as the nullability annotations
+   * of the type parameters for the LHS expression
```
```suggestion
   * For conditional expressions, check whether the type parameters for the sub-expressions of
   * the conditional expression have the same nullability annotations as the nullability annotations
   * of the type parameters for the conditional expression itself
```

msridhar reviewed (commented):
Just one minor comment on the docs and then this should be good to go!

NikitaAware committed:
Update nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

Co-authored-by: Manu Sridharan <msridhar@gmail.com>

msridhar committed:
formatting

msridhar committed:
clarify docs and add test

msridhar reviewed (approved):
Looks good to me!  We can also open a PR for this one on the main NullAway repo

nimakarimipour committed:
Refactoring in symbol serialization (#736)

This PR refactors the serialization logic to call a `static` method in `Serializer` to convert a `Symbol` to its corresponding `String` representation. This PR is in preparation for the update on serialization service in #735.

NikitaAware committed:
Add JSpecify checking for return statements (#734)

For JSpecify, this pull request adds a check that the type parameter nullability of the return type of the method matches the type of any returned expression.  For example:
```
static A<@Nullable String> test() {
  return new A<String>();
}
```
For the method return type, the type parameter is `@Nullable String`, but the type of the type parameter of the returned expression is `@NonNull String`. NullAway will now report an error here.

nimakarimipour committed:
Refactoring tabSeparatedToString logic to prepare for serialization version 3. (#738)

This PR is merely refactoring which prepares serialization logic for version `v3` implemented in #735.

Changes in this PR

- Uses adapter to serialize method signatures. 
- Adds `SerializationAdapter` to `tabSeparatedToString` methods for symbol serialization.
- Adds `SerializationAdapter` to `serializeSymbol` static method for symbol serialization.

msridhar committed:
Merge branch 'master' into generics-ternary-operator-assignment-changes
